### ISSUE TITLE
Potential memory leak in Python binding for `simplify_pattern`

### ISSUE THREAD

**Issue by tothtamas28** [runtimeverification/llvm-backend#922]
Running `simplify_pattern` iteratively results in monotonically increasing memory consumption. 

[test_simplify.py.txt](https://github.com/runtimeverification/llvm-backend/files/13614189/test_simplify.py.txt)

Copy the file to `pyk` repo root, then run

```
poetry run python3 test_simplify.py
```

**Comment by Baltoli** [comment]
We fixed a couple of minor memory leaks in https://github.com/runtimeverification/llvm-backend/pull/924; I confirmed that this particular reproduction still happens after doing so.

**Comment by Baltoli** [comment]
I have created a smaller C reproduction of the issue that avoids Python entirely in the interests of making it easier to use traditional memory profiling tools; this more minimal reproduction still exhibits the same monotonic increase in memory usage. It needs a couple of paths to be set up but isn't too onerous to run.

`simplify.c`
```c
#include <kllvm-c/kllvm-c.h>
#include <stdio.h>
#include <stdlib.h>

int main() {
  kore_pattern *pat = kore_pattern_parse(
      "Lbl'UndsPlus'Int'Unds'{}(\\dv{SortInt{}}(\"1\"),\\dv{SortInt{}}(\"2\")"
      ")");

  kore_sort *sort = kore_composite_sort_new("SortInt");

  for (int i = 0;; ++i) {
    if (i % 10000 == 0) {
      fprintf(stderr, ".");
    }

    char *data;
    size_t size;
    kore_simplify(pat, sort, &data, &size);
    free(data);
  }

  kore_sort_free(sort);
  kore_pattern_free(pat);
```

**Comment by Baltoli** [comment]
Rewrite step reproduction; from the Pyk root as usual:

`src/tests/integration/k-files/count.k`
```k
module COUNT
  imports INT
  imports BOOL
  
  syntax C ::= count(Int)
  rule count(I) => count(I -Int 1) requires I =/=Int 0
  rule count(I) => true requires I ==Int 0
endmodule
```

`init.kore`
```
Lbl'-LT-'generatedTop'-GT-'{}(Lbl'-LT-'k'-GT-'{}(kseq{}(inj{SortC{},SortKItem{}}(Lblcount'LParUndsRParUnds'COUNT'Unds'C'Unds'Int{}(\dv{SortInt{}}("1000000"))),dotk{}())),Lbl'-LT-'generatedCounter'-GT-'{}(\dv{SortInt{}}("0")))
```

`test_steps.py`
```python
import logging
from itertools import chain, cycle, repeat
from pathlib import Path
from tempfile import TemporaryDirectory

from pyk.kllvm import compiler, importer
from pyk.ktool.kompile import kompile


MAIN_FILE = 'src/tests/integration/k-files/count.k'
DOT_PERIOD = 100


logging.basicConfig(level=logging.INFO)

with TemporaryDirectory() as tmp:
    definition_dir = Path(tmp) / 'count-kompiled'
    kompile(MAIN_FILE, output_dir=definition_dir)
    compiler.compile_kllvm(definition_dir)
    compiler.compile_runtime(definition_dir)

    importer.import_kllvm(definition_dir)
    from pyk.kllvm import parser

    runtime = importer.import_runtime(definition_dir)
    runtime.init_static() # comment out to see leakage!

    dot = cycle(chain(repeat(False, DOT_PERIOD - 1), (True,)))
    while True:
        if next(dot):
            print('.', end='', flush=True)

        init_pattern = parser.parse_pattern_file('init.kore')
        term = runtime.term(init_pattern)
        term.step(-1)
```

Reproduction:
```console
$ poetry run python3 test_steps.py
```

**Comment by Baltoli** [comment]
Closed by #945 and https://github.com/runtimeverification/pyk/pull/798
